packages/components/src/circular-option-picker/circular-option-picker.tsx

```diff
@@ -85,38 +84,25 @@ function ListboxCircularOptionPicker(
 		children,
 		...additionalProps
 	} = props;
-	const rtl = isRTL();
 
-	const compositeState = useCompositeState( { baseId, loop, rtl } );
-	const { setBaseId, setLoop, setRTL } = compositeState;
-
-	// These are necessary as `useCompositeState` is sealed after
-	// the first render, so although unlikely to happen, if a state
-	// property should change, we need to process it accordingly.
-
-	useEffect( () => {
-		setBaseId( baseId );
-	}, [ setBaseId, baseId ] );
-
-	useEffect( () => {
-		setLoop( loop );
-	}, [ setLoop, loop ] );
-
-	useEffect( () => {
-		setRTL( rtl );
-	}, [ setRTL, rtl ] );
+	const compositeStore = useCompositeStore( {
+		focusLoop: loop,
+		rtl: isRTL(),
+	} );
 
 	const compositeContext = {
 		isComposite: true,
-		...compositeState,
+		baseId,
+		...compositeStore,
 	};
 
 	return (
 		<div className={ className }>
 			<CircularOptionPickerContext.Provider value={ compositeContext }>
```

Comment:
Not sure how I feel about this now. `ariakit` provides a `useCompositeStore` hook, so it's not strictly necessary to have our own provider sharing the composite store, as any children can get that easily enough. However we would ideally keep `compositeContext` around to share state not tied to `Composite` store data. Left as is for now to keep as close to original implementation as possible.

Reply:
This isn't a big issue, but I'd recommend not spreading the store object into a new, unmemoized one. Doing so negates the advantages of referential stability. The `compositeStore` object maintains enough stability to use in effect dependencies safely.

A simple fix would be to `useMemo` here:

```jsx
const compositeContext = React.useMemo(() => ({
  isComposite: true,
  baseId,
  ...compositeStore,
}), [baseId, compositeStore]);
```

This will also reduce the amount of times the context consumers re-render unnecessarily.

Reply:
> Alternatively, you can pass the store object to a property, instead of spreading:

This is probably a cleaner approach, as it will also help with code readability when de-structuring the context

Reply:
I've simplified this a bit, removing `isComposite` from `compositeContext`, so it now just has an (optional) `baseId` and `compositeStore`...

```ts
type CompositeContext = {
  baseId?: string;
  compositeStore?: CompositeStore;
};
```

To both your points, this solves multiple problems, and makes things a fair bit clearer.